### ISSUE TITLE
Translated field's value change doesn't get reflected in Translated Term's Internal Source for Odoo's default language other than English (US)

### ISSUE THREAD

**Issue by blackwaremayank** [odoo/odoo#44357]
Impacted versions: 13

Steps to reproduce:

1. Create new Odoo Enterprise Database with Language **English (CA)**

2. Now install **Sales Management** module and also enable **French (CA)** language.

3. Now create one Product and Save it.
![image](https://user-images.githubusercontent.com/48239869/73523266-7b8a4680-4430-11ea-8882-b7fff7091425.png)

4. Now edit this same product and change the name of the product.
![image](https://user-images.githubusercontent.com/48239869/73523412-dcb21a00-4430-11ea-9851-704c8e9b5f36.png)

5. Prepare one Sale Order like shown in below snapshot.
![image](https://user-images.githubusercontent.com/48239869/73523747-8db8b480-4431-11ea-9a2d-b7ad97569839.png)

6. Facing the same issue when i create new Purchase Order as shown in below snapshot.
![image](https://user-images.githubusercontent.com/48239869/73523822-b771db80-4431-11ea-9a27-59266d56b72d.png)

Current behavior:

- Product's Name isn't updated properly as **Internal Source** Translated Terms.

Expected behavior:

- Product's Name should be updated properly as **Internal Source** of Translated Terms.

![image](https://user-images.githubusercontent.com/48239869/73524266-c7d68600-4432-11ea-806a-148ad1eb4ff3.png)

Video/Screenshot link (optional):

**Note ** This's general issue with all translated fields of the system.

**Comment by Yenthe666** [comment]
@mart-e

**Comment by pedrobaeza** [comment]
The product is translated to the partner's language, so that's the problem here.

**Comment by intero-chz** [comment]
@pedrobaeza he is expecting that language "English (CA)" is the default language of his Odoo, so changing the product's name should also change the source term.

That was never the case in Odoo IIRC and IMO one of Odoo's biggest translation problems for customers using another language than English (US) as default language.

So it's not a bug but a feature, a hard to understand one for customers.

**Comment by pedrobaeza** [comment]
OK, then clarified, but indeed that's not happening, and honestly I don't see either why should it goes that way. Let's see what @mart-e thinks about this.

**Comment by Yenthe666** [comment]
By default I believe it was said that you should then install English as a primary, then install French and create the terms in English in order to translate them to French, no?

**Comment by intero-chz** [comment]
@pedrobaeza technical i can understand the way Odoo is doing it. English (US) is default for code and every other thing.

For a user that is hard to understand. And using more than one language (we have one customer with 8 languages) is like pushing efforts exponentionally when changing like source terms or other translation related things. And not using English (US) as default language (just normal for German customers) doesn't make it easier :-(

**Comment by pedrobaeza** [comment]
I'm not using English as default language as well (I'm Spanish), and I don't find this kind of problems. The key is to always provide proper `lang` in context. Maybe you have any automatism that don't do that? The example provided here is a convoluted one because you have to select a partner. Not doing so, although is tolerated, tend to fail in lot of things (prices, language, etc), so it's not a problem per se of language handling. In previous versions, you weren't able to add a line without selecting a partner.

**Comment by intero-chz** [comment]
> By default I believe it was said that you should then install English as a primary, then install French and create the terms in English in order to translate them to French, no?

And that is the thing customers don't want to understand. It means more costs for them. Why should for example a German company with german-only based business cases have english translations in Odoo?

**Comment by mart-e** [comment]
Hi,

> English (US) is default for code and every other thing

In 13.0, I have tried to unify a bit the behaviour by generating en_US translations too when you are in multilanguage env (before translations were only used for other languages) but this is not the issue here.
The problem is because we indeed try to browse the product with the customer language (being `None`) so we bypass all the translation mechanism.

https://github.com/odoo/odoo/blob/5d593eb31ce7d545a57cb93bc9834e4cebadea19/addons/sale/models/sale.py#L1431-L1432

I think this code can be improved and we can fallback on the language of the company.
In 13.0, we have added a method `get_lang` to do this kind of things. This method will make sure to always return a language.

https://github.com/odoo/odoo/blob/5d593eb31ce7d545a57cb93bc9834e4cebadea19/odoo/tools/misc.py#L1167-L1178

My personal opinion is that we should never have `lang=None` anywhere in the code but the ORM.
I haven't convinced (yet) higher instances to make the language required so this is strictly my opinion...
